src/Formats/JSONUtils.cpp

```diff
@@ -286,11 +291,78 @@ namespace JSONUtils
+            auto do_deserialize = [](IColumn & column_, ReadBuffer & buf_, auto && check_for_empty_string, auto && deserialize) -> bool
+            {
+                if (check_for_empty_string(buf_))
+                {
+                    column_.insertDefault();
+                    return false;
+                }
+                else
+                    return deserialize(column_, buf_);
+            };
+
+            auto deserialize_impl = [as_nullable, &format_settings, &serialization](IColumn & column_, ReadBuffer & buf_) -> bool
+            {
+                if (as_nullable)
+                    return SerializationNullable::deserializeNullAsDefaultOrNestedTextJSON(column_, buf_, format_settings, serialization);
+
+                serialization->deserializeTextJSON(column_, buf_, format_settings);
+                return true;
+            };
+
+            if (!format_settings.json.empty_as_default  || in.eof() || *in.position() != EMPTY_STRING[0])
+                return do_deserialize(column, in, [](ReadBuffer &) { return false; }, deserialize_impl);
+
+            if (in.available() >= EMPTY_STRING_LENGTH)
+            {
+                /// We have enough data in buffer to check if we have an empty string.
+                auto check_for_empty_string = [](ReadBuffer & buf_)
+                {
+                    auto * pos = buf_.position();
+                    if (checkString(EMPTY_STRING, buf_))
+                        return true;
+                    else
+                    {
+                        buf_.position() = pos;
+                        return false;
+                    }
+                };
+
+                return do_deserialize(column, in, check_for_empty_string, deserialize_impl);
+            }
+
+            /// We don't have enough data in buffer to check if we have an empty string.
+            /// Use PeekableReadBuffer to make a checkpoint before checking for an
+            /// empty string and rollback if check was failed.
+
+            auto check_for_empty_string = [](ReadBuffer & buf_) -> bool
+            {
+                auto & peekable_buf = assert_cast<PeekableReadBuffer &>(buf_);
+                peekable_buf.setCheckpoint();
+                SCOPE_EXIT(peekable_buf.dropCheckpoint());
+                if (checkString(EMPTY_STRING, peekable_buf))
+                    return true;
+                else
+                {
+                    peekable_buf.rollbackToCheckpoint();
+                    return false;
+                }
+            };
+
+            auto deserialize_impl_with_check = [&deserialize_impl](IColumn & column_, ReadBuffer & buf_) -> bool
+            {
+                auto & peekable_buf = assert_cast<PeekableReadBuffer &>(buf_);
+
+                if (!deserialize_impl(column_, peekable_buf))
+                    return false;
+                if (likely(!peekable_buf.hasUnreadData()))
+                    return true;
+                return false;
```

Comment:
It's a bit incorrect. We should check `peekable_buf.hasUnreadData()` in any return value from `deserialize_impl`. And when `peekable_buf.hasUnreadData()` we should throw an exception, because it means that the state is incorrect and we cannot continue parsing. Actually, in our case this should never happen, so we can even throw `ErrorCodes::LOGICAL_ERROR`:
```suggestion
                bool res = deserialize_impl(column_, peekable_buf));
                if (unlikely(peekable_buf.hasUnreadData()))
                    throw Exception(ErrorCodes::LOGICAL_ERROR, "Incorrect state while parsing JSON: PeekableReadBuffer has unread data in own memory: {}", String(peekable_buf.position(), peekable_buf.available()));
                return false;
```